keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md

```diff
@@ -0,0 +1,912 @@
+  // always override counters from mixins.
+  //
+  // The mixins referenced here must be defined in the same
+  // ResourceSlice.
+  //
+  // The maximum number of includes is 8.
+  //
+  // +featureGate=DRAResourceSliceMixins
+  // +optional
+  // +listType=atomic
+  Includes []string
+}
+```
+
+### Implementation
+
+The DRA scheduler will keep the mixin structure throughout the scheduling process
+as much as possible and avoid completely flattening the ResourceSlices. This is
+to avoid additional memory usage that might come as a result. For example, we
+plan to walk the mixins as part of the CEL variable lookup to avoid having to
+flatten the device representation.
+
+If the mixins feature is disabled, any devices or counter sets that
+references mixins will be droppped. This also means that all devices that references
+a dropped counter set will also be dropped. The result is that the scheduler will not
+see those devices. From the users point of view, the consequence is that the scheduler
+might pick a different device than expected or fails to allocate any device at all. But we
+think this failure mode is preferable than allowing the scheduler to make allocation decisions
+based on incomplete data.
+
+### Limits
+
+For DRA, we have gradually moved away from individual per-slice and per-map limits towards
+aggregating at the higher level. The reason for this is to give users maximum flexibility between
+defining a small number of complex devices or a large number of simple devices in a single
+ResourceSlice without exceeding the limit on the size of Kubernetes objects.
+
+For this KEP, we propose taking this to what is essentially the logical conclusion, where
+we enforce most of the limits across all devices, mixins and counter sets in a ResourceSlice,
+rather than setting separate limits for each of them.
+
+The ResourceSlice-wide limits will be:
+* Total number of devices is 128.
+* Total combined number of attributes and capacity in a ResourceSlice is 4096 (so with the maximum number of devices, there can be 32 per device).
+* Total number of counters is 256.
+* Total number of consumed counters is 2048 (so with the maximum number of devices, there can be 16 per device).
+
+We will still enforce some per-slice limits:
+* The number of mixins that can be referenced from each device, counter set or device counter consumption is 8.
+* The number of taints per device is 4.
+
+We will also enforce one limit on the flattened device:
+* The combined number of attributes and capacities for a single device can not exceed 32. We do this
+  to avoid increasing the cost of evaluation the CEL expressions for a device.
+
+The limits on the number of counters across counter sets, mixins and device counter consumption in 1.33 for the
+Partitionable Devices KEP will be removed, as those are still in alpha.
+The limit of 32 on the number of attributes and capacities per device will be removed over the next 2 releases (1.34 and 1.35) to
```

Comment:
I'm not sure about this. The problem I mentioned in https://github.com/kubernetes/enhancements/pull/5238#discussion_r2062707393 ("makes the estimated cost of the same expression higher than it was before" and therefore invalid) remains, even if we do this gradually. We would have to increase the cost limit, but don't know by how much.

Reply:
Yeah, I've updated the language here. I think we essentially just want to change the validation to do the check with mixins applied rather than on just the device. That should be a safe change I think, since all devices that fail the current check will also fail the updated one.